### PR TITLE
Limit private repos

### LINKED ISSUE #3653: Limit maximum count of private or public repositories for individual users
- Gitea version (or commit ref): 1.4.0+rc1-17-g93f34fd
- Git version: 
- Operating system: Ubuntu 17.10
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description

Hello,

At the moment it is possible to select the maximum count of repositories a user is allowed to create.

Can the admin specify the count for public or private repositories individually? What about managing how many forks, mirrors, etc... ?

If the admin cant do this right now would this be a good feature?

Thank you,

Adrian

### PR DESCRIPTION
**Pull request by mvdkleijn**
Replacement PR for #12705 since I done f-ed up that one.

**Note:** this turned out a *little* bigger than I had anticipated... as I stated before, I'm completely new to the Gitea codebase so I probably missed a lot.

Also, I didn't take a look at the tests yet. I did do some basic manual tests and as far as I can tell, the basic functionality is working.

Anyone who's willing to help test this, help fix tests and/or review it for missed parts or improvements, please comment. (I turned on "allow edits by maintainers" for this PR by the way)

**Original description**

Since @kasbah does not appear to have time to continue his work on PR #6184 I started looking at continuing his work. As there have been many changes to master since, I felt it was easier to start fresh.

Fixes #3653. Feedback very much appreciated.

### CODE REVIEW

**Comment by mvdkleijn** [comment]
Apologies for the confusion and overhead by the way. :smile:

**Comment by 6543** [comment]
A forcepush after a rebase would have done it too ... but here we are ;)

**Comment by 6543** [comment]
@mvdkleijn can you commit the result of `make generate-swagger` ?

**Comment by mvdkleijn** [comment]
> A forcepush after a rebase would have done it too ... but here we are ;)

Which is exactly what I **thought** I did but that's the part where things went wrong. Anyway, it doesn't matter. I'm happier with this PR anyway. :smile:

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -2012,8 +2013,10 @@ users.auth_login_name = Authentication Sign-In Name
 users.password_helper = Leave the password empty to keep it unchanged.
 users.update_profile_success = The user account has been updated.
 users.edit_account = Edit User Account
-users.max_repo_creation = Maximum Number of Repositories
-users.max_repo_creation_desc = (Enter -1 to use the global default limit.)
+users.max_public_repo_creation = Maximum Number of Public Repositories
+users.max_public_repo_creation_desc = (Enter -1 to use the global default limit.)
+users.max_private_repo_creation = Maximum Number of Private Repositories
+users.max_private_repo_creation_desc = (Enter -1 to use the global default limit.)
```
Comment (bagasme) [inline options/locale/locale_en-US.ini:2019]: ack

Comment (mvdkleijn) [reply options/locale/locale_en-US.ini:2019]: I suppose that means "this violates DRY" or some such? :smile:

**Comment by mvdkleijn** [comment]
I have a question. The system currently keeps track of the number of repos that a user owns by storing it as a field in the DB. It causes a lot of code snippets all over the codebase. Was that done for performance reasons or otherwise? It feels like it could be a nice improvement in terms of maintenance to simply SELECT COUNT entries in the DB instead of storing the counts in the DB as a field.

**Comment by zeripath** [comment]
Yup it's an optimisation for performance reasons.

**Comment by mvdkleijn** [comment]
Just checking in... my four year old gave me the gift of the flu. I'll continue on this in a few days.

**Pull request closed without merging**